File: problems/0109_convert-sorted-list-to-binary-search-tree/solution_inorder_sim.cpp
class Solution {
  private:
    // Walks the list in original order while the recursion claims nodes
    // exactly where an inorder insertion would place them.
    ListNode *current = nullptr;

  public:
    TreeNode *sortedListToBST(ListNode *head) {
        // One sizing pass first: the recursion needs each subtree's node
        // count to pick the same middles the midpoint walk would.
        int count = 0;
        for (ListNode *node = head; node != nullptr; node = node->next) {
            count++;
        }
        current = head;
        return build(0, count);
    }

  private:
    TreeNode *build(int lo, int hi) {
        if (lo >= hi) {
            return nullptr;
        }
        // The left subtree is the first half of [lo, hi) — the same
        // tie-break as the midpoint walk, so both variants build the
        // identical tree.
        int mid = (lo + hi) / 2;
        TreeNode *left = build(lo, mid);
        // Inorder position: after the left subtree, the next node in
        // original order is the root; the cursor hands it over and steps
        // forward, then the right subtree takes what remains.
        TreeNode *root = new TreeNode(current->val);
        current = current->next;
        root->left = left;
        root->right = build(mid + 1, hi);
        return root;
    }
};

File: problems/0100-0200/0109_convert-sorted-list-to-binary-search-tree/solution_inorder_sim.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        # One sizing pass first: the recursion needs each subtree's node
        # count to pick the same middles the midpoint walk would.
        count = 0
        node = head
        while node is not None:
            count += 1
            node = node.next
        # The cursor walks the list in original order; the recursion claims
        # nodes exactly where an inorder insertion would place them.
        current = head

        def build(lo: int, hi: int) -> Optional[TreeNode]:
            nonlocal current
            if lo >= hi:
                return None
            # The left subtree is the first half of [lo, hi) — the same
            # tie-break as the midpoint walk, so both variants build the
            # identical tree.
            mid = (lo + hi) // 2
            left = build(lo, mid)
            # Inorder position: after the left subtree, the next node in
            # original order is the root; the cursor hands it over and
            # steps forward, then the right subtree takes what remains.
            root = TreeNode(current.val)
            current = current.next
            root.left = left
            root.right = build(mid + 1, hi)
            return root

        return build(0, count)
